sdk-cpp/src/internal/com/do_download_property_internal.cpp

```diff
@@ -0,0 +1,160 @@
+
+
+#include "do_download_property_internal.h"
+
+#include <cassert>
+#include <vector>
+#include <string>
+
+#include "do_exceptions.h"
+
+using namespace microsoft::deliveryoptimization::details;
+
+int32_t UTF8toWstr(std::wstring& wstr, const char* str, size_t cch = 0)
+{
+    if (cch == 0)
+    {
+        cch = strlen(str);
+    }
+
+    if (cch == 0)
+    {
+        wstr = std::wstring();
+    }
+
+    std::vector<wchar_t> dest(cch * 4);
+    const uint32_t result = MultiByteToWideChar(CP_UTF8, 0, str, static_cast<int>(cch), dest.data(), static_cast<int>(dest.size()));
+    if (result == 0)
+    {
+        return E_FAIL;
+    }
+    wstr = std::wstring(dest.data(), result);
+    return S_OK;
+}
+
+CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(const std::string& val)
+{
+    V_VT(&_var) = VT_BSTR;
+
+    std::wstring wval;
+    const auto res = UTF8toWstr(wval, val.c_str());
+#if (!DO_DISABLE_EXCEPTIONS)
+    throw_if_fail(res);
+#endif
+    _initCode = res;
+
+    BSTR bstr = SysAllocString(wval.c_str());
+    if (bstr == nullptr)
+    {
+ #if (!DO_DISABLE_EXCEPTIONS)
+        throw_if_fail(DO_ERROR_FROM_STD_ERROR(std::errc::not_enough_memory)); // empty bstring may not always be out of memory error, but should be most of the time
+#endif
+        _initCode = DO_ERROR_FROM_STD_ERROR(std::errc::not_enough_memory);
+    }
+    V_BSTR(&_var) = bstr;
+};
+
+CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(uint32_t val)
+{
+    V_VT(&_var) = VT_UI4;
+    V_UI4(&_var) = val;
+};
+
+CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(uint64_t val)
+{
+    V_VT(&_var) = VT_UI8;
+    V_UI8(&_var) = val;
+};
+
+CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(bool val)
+{
+    V_VT(&_var) = VT_BOOL;
+    V_BOOL(&_var) = val ? VARIANT_TRUE : VARIANT_FALSE;
+};
+
+CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(std::vector<unsigned char>& val)
+{
+    throw errc::e_not_impl;
+};
+
+CDownloadPropertyValueInternal::CDownloadPropertyValueInternal(const download_property_value::status_callback_t& val)
+{
+    _callback = val;
+}
+
+int32_t CDownloadPropertyValueInternal::init_code() const noexcept
+{
+    return _initCode;
+}
+
+CDownloadPropertyValueInternal::~CDownloadPropertyValueInternal()
+{
+#ifdef DEBUG
+    //TODO(jimson): Variant clear fails with DISP_E_BADVARTYPE when used for callbacks, so the assertion will terminate the application
```

Comment:
>DISP_E_BADVARTYPE [](http://example.com/codeflow?start=45&length=17)

This is most likely because VariantInit is not done. See line 82 above.

Reply:
Ah yes, good catch